### ISSUE TITLE
feat: add assets trimming to/in fluent blazor wasm

### ISSUE THREAD

**Issue by baywet** [microsoft/fast-blazor#239]
# 🙋 Feature Request

## 🤔 Expected Behavior

The service injection should allow people to specify what they need.

## 😯 Current Behavior

Related #225 
Not being able to trim the assets introduced by Fluent Blazor in a WASM SPA app has multiple downsides:
- It increases deployment length (downtime, failures risk, CD costs...)
- **It increases assets service workers size (from a couple of KBs to 2.5 MBs)**
- It decreases performance and user satisfaction with the app

All those issues reduce my incentive of using these components as opposed to other design languages.

## 💁 Possible Solution

```CSharp
builder.Services.AddFluentUIComponents();
```
Would bring everything like it does today.

```CSharp
builder.Services.AddFluentUIComponents(x => x.AddAll24Icons());
```

Would add all icons of size 24

```CSharp
builder.Services.AddFluentUIComponents(x => x.AddAllLozalizedIcons("fr"));
```

Would add all icons localizations for the French locale

```CSharp
builder.Services.AddFluentUIComponents(x => x.AddIcons(FluentIcons.Home));
```

Would add all the Home icons variants

```CSharp
builder.Services.AddFluentUIComponents(x => x.AddIconsFor<FluentFlipper>());
```

Would add all basic icons for the flipper component

```CSharp
builder.Services.AddFluentUIComponents(x => x.AddIconsFor<FluentFlipper>()
                                                                             .AddAll24Icons()
                                                                             .AddIcons(FluentIcons.Home)
                                                                             .AddAllLozalizedIcons("fr")
);
```

Would add a union of all the sets mentioned above.

## 🔦 Context

I've build a blazor wasm app, added fluent UI blazor, and published the app

```Shell
dotnet new blazorwasm
dotnet add package Microsoft.Fast.Components.FluentUI --prerelease
```

Added `builder.Services.AddFluentUIComponents()` in Program.cs
Added `<ServiceWorkerAssetsManifest>service-worker-assets.js</ServiceWorkerAssetsManifest>` within the `PropertyGroup` tag.
Added `<ServiceWorker Include="wwwroot\service-worker.js" PublishedContent="wwwroot\service-worker.published.js" />` with a `ItemGroup` tag.

```Shell
dotnet publish -c Release
```

The output service worker is now 2.5MBs and the content folder is now 7MBs.

## 💻 Examples

CC @jeepNL

**Comment by JeepNL** [comment]
I know MudBlazor had a similar issue with trimming the svg icons. More about this here: https://github.com/MudBlazor/MudBlazor/pull/4639

and here: https://github.com/MudBlazor/MudBlazor/issues/2424

They've found a solution, actually MudBlazor is now very small after publish. Less than 400kb Brotli compressed.

**Comment by vnbaaij** [comment]
I would need to take a look at why the `ServiceWorker` grows to 2.5 Mb. 

The size of the content/wwwroot folder is expected as there are about 12K icons (as SVG files). The way we handle the icons is quite different than the way it is done in MudBlazor. On our side they are added as regular assets in the `wwwroot` folder of the project. As the library is an RCL, they are treated like ordinary static files on the server. They don’t get downloaded until requested so no need to trim anything there.

**Comment by baywet** [comment]
Asides from the service worker for which you need to look at. Having to deploy that many assets that are not going to be used at each deployment is a tax on the CD + storage costs. Having the ability to narrow down the scope would go a long way.

**Comment by vnbaaij** [comment]
> Having to deploy that many assets that are not going to be used at each deployment is a tax on the CD 

Not really. The deployment is done automatically as part of the `dotnet publish`. No extra steps needed.

> storage costs. 

Well, the whole `wwwroot` folder is less than 7Mb so can't imagine that being a really big issue nowadays.

We have no way of knowing which assets are going to be used. If we would us a font, we would also include all icons. Downside there is that the whole font would need to be downloaded (couple of hundred KBs) even if you would use only 1 icon

**Comment by baywet** [comment]
> Not really. The deployment is done automatically as part of the dotnet publish. No extra steps needed.

Are you suggesting that dotnet publish is able to upload the results straight to Azure Storage Static Website? because I wasn't able to find documentation for that in my initial search. It's either using an [additional task](https://github.com/Azure/static-web-apps-deploy) to deploy to static web app (not what I'm using) or leveraging the Azure CLI (or equivalent) to upload to the storage container.

**Comment by vnbaaij** [comment]
Yes, sorry. You are right. Poorly woorded from my site. Dotnet publish makes sure all the stuff is in the right structure but it must indeed still be uploaded. Static web apps makes this so easy, you tend to forget...

**Comment by baywet** [comment]
no worries. Just to put us back on track about what's left on this issue:
- You're going to have a look at why the assets service worker balloons up
- depending on the outcome of that investigation, providing more control to people over what gets emitted might be necessary

And to be clear, I wasn't suggesting any changes to the structure of what's being emitted (embedding the fonts or not etc...) but simply to provide some kind of "filtering" mechanism over what's being emitted to app developers so they can fine tune things.

**Comment by vnbaaij** [comment]
> You're going to have a look at why the assets service worker balloons up

Yes. Although I don't have much experience with PWAs/Service workers but I'll see what I can find

> depending on the outcome of that investigation, providing more control to people over what gets emitted might be necessary

Yes.

**Comment by baywet** [comment]
no worries, FYI here are the steps I followed to make a PWA from my blazor wasm app https://learn.microsoft.com/en-us/aspnet/core/blazor/progressive-web-app?view=aspnetcore-6.0&tabs=visual-studio#convert-an-existing-blazor-webassembly-app-into-a-pwa

And the source lives here in case you can repro locally in a new app and you want to poke around with ours. https://github.com/microsoft/kiota/tree/main/src/Kiota.Web

**Comment by baywet** [comment]
Note: in addition to your investigation, one of the solutions might be to provide clear control guidance for the assets that are being cached https://learn.microsoft.com/en-us/aspnet/core/blazor/progressive-web-app?view=aspnetcore-6.0&tabs=visual-studio#control-asset-caching

**Comment by vnbaaij** [comment]
When I add below settings to the `.csproj` of the library, I get all kind of errors (mostly on `System.Text.Json`) when the site is deployed to Azure Static Web Apps. Any idea on what is needed from there and what not? And what might be the cause of the errors?
```
<IsTrimmable>true</IsTrimmable>
<PublishTrimmed>true</PublishTrimmed>
<TrimMode>link</TrimMode>
<EnableTrimAnalyzer>true</EnableTrimAnalyzer>
```

For now, these settings have been removed from the source.

**Comment by baywet** [comment]
Not sure, I haven't played with that part too much. It's probably worth reaching out to the upstream dependency (dotnet runtime?)

**Comment by vnbaaij** [comment]
I tried excluding the svg files by adding a regular expression to the `offlineAssestsExclude` line in the `service-worker.pubished.js` file. However, the svg files are stlll being included in the `wwwroot` folder and, more importantly, in the `service-worker-assets` file.
I'm asking around internally if there is a solution for this.

**Comment by vnbaaij** [comment]
Hi, 
No progress on the ballooning of the service worker, but...I do have something in the works where you as the developer can determine what icons and emojis (new component) assets get deployed when publishing the app. So essentially offering you trimming capabilities. Please see the discussion on this: https://github.com/microsoft/fast-blazor/discussions/300

**Comment by baywet** [comment]
Thanks for making progress on this. At least deployments will be much faster with trimmed icon sets. Am I correct to assume that trimming the icons won't impact the worker size at this point?

**Comment by vnbaaij** [comment]
I haven't tried it yet, but I'm guessing the worker size is still going to be an issue. To solve that, we would probably need to write custom logic for the MSBuild phase. And if we want to control what gets cached we need to modify the logic in onInstall in service-worker.published.js as described here: https://learn.microsoft.com/aspnet/core/blazor/progressive-web-app#background-updates. It's on the ToDo list...

**Comment by vnbaaij** [comment]
Closed as (mostly) done in 2.1 release
